docs/proposals/devfile/outer-loop-build-and-deploy.md

```diff
@@ -0,0 +1,438 @@
+
+`location`: Dockerfile location which can be an URL or a path relative to buildContext
+
+`args`: Argument list for the docker build (optional)
+
+`rootRequired`: Specifies whether a privileged builder pod is required.  Default is false. (optional)
+
+__Note:__
+1. A common pattern will be using a global variable to define the image name (`myimage` in the example above) so that it can be easily referred to in the deploy step later.
+2. The `apply` command with the `image` component is optional. If the `apply` command that references an image component does not exist, then the image build will be done at the startup automatically (similar to the behaviour of the existing `kubernetes` components).  If there is an `apply` command that references the `image` component, then the user will need to create a composite command to chain the image build apply command with the deploy command (see the deployment command section) to complete the build and deploy.
+
+### Example using a file Dockerfile with registry and secret for the image push:
+    variables:
+    myimage: myimagename
+    components:
+    - name: mydockerfileimage
+        image:
+        imageName: {{myimage}}
+        dockerfile:
+            buildContext: ${PROJECTS_ROOT}/build
+            location: https://github.com/redhat-developer/devfile-sample/blob/master/src/Dockerfile
+            args: [ "arg1", "arg2", "arg3" ]
+            rootRequired: false
+            envFrom:
+            - secretRef:
+                name: my-secret
+
+`secretRef`: The reference to the secret for pushing the image to the registry
+
+For the secrets, it should support the same mechanisms as specified in https://github.com/devfile/api/issues/299.
+
+### Example using `apply` command on an image component:
+    commands:
+    - id: deploybuild
+        apply:
+        component: mydockerfileimage
+
+### Example using a Dockerfile stored within the devfile registry as a resource with `apply` command:
+    variables:
+    myimage: myimagename
+    components:
+    - name: mydockerfileimage
+        image:
+        imageName: {{myimage}}
+        dockerfile:
+            id: mycompany/my-node-stack-dockerfile/v2.2.2
+            args: [ ]
+    commands:
+    - id: deploybuild
+        apply:
+        component: mydockerfileimage
+ 
+### Example using a Dockerfile stored in a git repo with push registry without apply command:
+    variables:
+    myimage: myimagename
+    components:
+    - name: mydockerfileimage
+        image:
+        imageName: {{myimage}}
+        dockerfile:
+            git:
+            remotes:
+                origin: "https://github.com/odo-devfiles/nodejs-ex.git"
+            location: Dockerfile
+            args: [ ]
+
+The git definition will be the same as the one in `starterProjects` definition that supports `checkoutFrom`
+
+`location`: the location of the dockerfile within the repo. If `checkoutFrom` is being used, the location will be relative to the `checkoutFrom`.
```

Comment:
This seems unclear to me -- `checkoutFrom` refers to a remote, so what does "relative" mean in this context?